Sources/GRPCCore/Transport/RetryThrottle.swift

```diff
@@ -0,0 +1,124 @@
+  /// The retry threshold, multiplied by 1000. If ``scaledTokensAvailable`` is above this then
+  /// retries are permitted.
+  private let scaledRetryThreshold: Int
+
+  /// Returns the throttling token ratio.
+  ///
+  /// The number of tokens held by the throttle is incremented by this value for each successful
+  /// response. In the context of throttling, a successful response is one which:
+  /// - receives metadata from the server, or
+  /// - is terminated with a non-retryable or fatal status code.
+  ///
+  /// If the response is a pushback response then it is not considered to be successful, even if
+  /// either of the preceding conditions are met.
+  public var tokenRatio: Double {
+    Double(self.scaledTokenRatio) / 1000
+  }
+
+  /// The maximum number of tokens the throttle may hold.
+  public var maximumTokens: Int {
+    self.scaledMaximumTokens / 1000
+  }
+
+  /// The number of tokens the throttle currently has.
+  ///
+  /// If this value is less than or equal to half of ``maximumTokens`` then RPCs will not be retried
+  /// and hedging will be disabled.
+  public var tokens: Double {
+    self.scaledTokensAvailable.withLockedValue {
+      Double($0) / 1000
+    }
+  }
+
+  /// Returns whether retries and hedging are permitted at this time.
+  public var isRetryPermitted: Bool {
+    self.scaledTokensAvailable.withLockedValue {
+      $0 > self.scaledRetryThreshold
+    }
+  }
+
+  /// Create a new throttle.
+  ///
+  /// - Parameters:
+  ///   - maximumTokens: The maximum number of tokens available. Must be in the range `1...1000`.
+  ///   - tokenRatio: The number of tokens to increment the available tokens by for successful
+  ///       responses. See the documentation on this type for a description of what counts as a
+  ///       successful response. Note that only three decimal places are used from this value.
+  /// - Precondition: `maximumTokens` must be in the range `1...1000`.
+  /// - Precondition: `tokenRatio` must be `>= 0.001`.
+  public init(maximumTokens: Int, tokenRatio: Double) {
+    precondition(
+      (1 ... 1000).contains(maximumTokens),
+      "maximumTokens must be in the range 1...1000 (is \(maximumTokens))"
+    )
+
+    let scaledTokenRatio = Int(tokenRatio * 1000)
+    precondition(scaledTokenRatio > 0, "tokenRatio must be >= 0.001 (is \(tokenRatio))")
+
+    let scaledTokens = maximumTokens * 1000
+    self.scaledMaximumTokens = scaledTokens
+    self.scaledRetryThreshold = scaledTokens / 2
+    self.scaledTokenRatio = scaledTokenRatio
+    self.scaledTokensAvailable = LockedValueBox(scaledTokens)
+  }
+
+  /// Records a success, adding a token to the throttle.
+  @usableFromInline
+  func recordSuccess() {
+    self.scaledTokensAvailable.withLockedValue { value in
+      value = Swift.min(self.scaledMaximumTokens, value &+ self.scaledTokenRatio)
+    }
+  }
+
+  /// Records a failure, removing tokens from the throttle. Returns whether retries will now be
+  /// throttled.
```

Comment:
Nit, but I guess we prefer
```suggestion
  /// Records a failure, removing tokens from the throttle. 
  /// - Returns: Whether retries will now be throttled.
```